### ISSUE TITLE
Unexpected results while training CBM on CelebA

### ISSUE THREAD

**Issue by nishadsinghi** [mateoespinosa/cem#5]
Hey Mateo,

Thanks for your work (I really admire your papers on concept-based models 😁) and releasing your code!

I am trying to train Concept Bottleneck Models using this repository on CelebA. Following your approach, I am using 8 concepts, but I don't have any hidden concepts (so, all 8 concepts are used). I tried training independent and sequential concept bottleneck models. This is the config:

> trials: 1
> results_dir: /mnt/qb/work/bethge/bkr046/CEM/results/CelebA
> 
> 
> dataset: celeba
> root_dir: /mnt/qb/work/bethge/bkr046/DATASETS/celeba_torchvision #cem/data/
> image_size: 64
> num_classes: 1000
> batch_size: 512
> use_imbalance: True
> use_binary_vector_class: True
> num_concepts: 8
> label_binary_width: 1
> label_dataset_subsample: 12
> #num_hidden_concepts: 0
> selected_concepts: False
> num_workers: 8
> sampling_percent: 1
> test_subsampling: 1
> 
> 
> intervention_freq: 1
> intervention_batch_size: 1024
> intervention_policies:
>     - "group_random_no_prior"
> 
> competence_levels: [1, 0]
> incompetence_intervention_policies:
>     - "group_random_no_prior"
> 
> skip_repr_evaluation: True
> 
> shared_params:
>     top_k_accuracy: [3, 5, 10]
>     save_model: True
>     max_epochs: 200
>     patience: 15
>     emb_size: 16
>     extra_dims: 0
>     concept_loss_weight: 1
>     learning_rate: 0.005
>     weight_decay: 0.000004
>     weight_loss: False
>     c_extractor_arch: resnet34
>     optimizer: sgd
>     early_stopping_monitor: val_loss
>     early_stopping_mode: min
>     early_stopping_delta: 0.0
>     momentum: 0.9
>     sigmoidal_prob: False
> 
> runs:
>     - architecture: 'SequentialConceptBottleneckModel'
>       extra_name: "NoInterventionInTrainingNoHiddenConcepts"
>       sigmoidal_embedding: False
>       concat_prob: False
>       embedding_activation: "leakyrelu"
>       bool: False
>       extra_dims: 0
>       sigmoidal_extra_capacity: False
>       sigmoidal_prob: True
>       training_intervention_prob: 0

While doing this, I noticed a few unexpected things:
1) Even with 100% interventions, the accuracy of the independent model is less than 50%. Since all concepts are visible, and it's an independent model, the c2y component would have seen all possible concept combinations during training, and therefore should be able to perform well when all ground-truth concepts are provided during interventions.

> +---------------------------------------------------------------------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+
> |                                   Method                                  |  Task Accuracy  | Concept Accuracy |   Concept AUC   |   25% Int Acc   |   50% Int Acc   |   75% Int Acc   |   100% Int Acc  |
> +---------------------------------------------------------------------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+
> | IndependentConceptBottleneckModel | 0.2743 ± 0.0000 | 0.8237 ± 0.0000  | 0.8163 ± 0.0000 | 0.3009 ± 0.0000 | 0.3229 ± 0.0000 | 0.3457 ± 0.0000 | 0.3711 ± 0.0000 |
> |  SequentialConceptBottleneckModel | 0.2784 ± 0.0000 | 0.8237 ± 0.0000  | 0.8163 ± 0.0000 | 0.3107 ± 0.0000 | 0.3409 ± 0.0000 | 0.3584 ± 0.0000 | 0.3711 ± 0.0000 |
> +---------------------------------------------------------------------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+

2) The training accuracy of the independent model is also pretty low, which is concerning. This is from the 49th epoch (which looks like the last epoch to me):

> Epoch 49:  86%|████████▌ | 24/28 [00:01<00:00, 14.65it/s, loss=3.13, y_accuracy=0.381, y_top_3_accuracy=0.548, y_top_5_accuracy=0.619, y_top_10_accuracy=0.714, val_y_accuracy=0.384, val_y_top_3_accuracy=0.557, val_y_top_5_accuracy=0.629, val_y_top_10_accuracy=0.697]

3) The number of classes in the dataset is 230 instead of 256. I guess that is because some of the 256 possible classes never appear in the dataset. Did you see this as well?

Do you think these observations are expected? If yes, it would be really helpful if you could provide some intuition as to why.

**Comment by mateoespinosa** [comment]
Hey Nishad!

So happy to hear you enjoy our work and that you are using this codebase!

Thank you so much for opening this ticket and raising this issue. At first sight, these results seem expected given the default architecture for the c2y model as well as the nature of the CelebA task used in this code base. However, let me try and answer the specific questions you asked and we can take the discussion from here (I may be able to help debugging this deeper if we realize there is an issue somewhere in the pipeline):
1. This is not entirely unexpected as the CelebA task is a very hard task by construction and therefore we do not expect models to be able to achieve very high accuracies from the get-go (top-1 accuracies between 30% and 45% are about right for resnet34-ish models). Because of this, we also log top-{2,5,10} accuracies for this task. I say this is not unexpected because this CelebA task has a heavy class imbalance both at the task level and at the concept level. This makes it a very hard task to solve perfectly, as it depends on one being able to properly (1) identify all concepts, which can be unbalanced or hard to correctly predict even though we picked the most balanced ones, and (2) identify the correct label from the concepts, which needs a c2y model that is capable of learning a correct mapping from binary concepts to their decimal representation. On top of this, some concept labels in CelebA are heavily subjective, and therefore they add an extra level of difficulty in correctly predicting all 8 selected attributes when one considers the noise in the ground truth annotations themselves.
However, as you say, when training an independent model one would expect the concept-to-task model to be slightly easier to learn (and therefore should have very high performance when all concepts are intervened). To check whether the issue is on that side of the CBM, have you tested more complex architectures for the c2y model than the default single linear layer? I would suggest starting by looking into that as it could be the case that the c2y model does not have the proper capacity to correctly learn the mapping we wish to learn (something that is not necessarily the case when using, say, Hybrid-CBMs or CEMs as the input to the c2y model is much richer). An easy way to give this a shot is by playing around with the`c2y_layers` field to indicate the size of MLP layers to use for the `c2y` model. Other things worth looking into are how long a model is trained as well as the early stop conditions. Finally, you can possibly scrap more performance from any model in this task by weighting the task loss taking into account the class distribution, as task labels are far from balanced if I recall correctly.
2. Just to confirm, is this the accuracy of the training of the c2y model when training the IndependentCBM? If so, then this may be potential evidence that the issue is indeed on this model. If possible, try out different architectures/hyperparameters. Because of the task-inherent difficulties discussed above, I don't expect this accuracy to be able to be improved significantly more than what you get here but you may obtain some gains with more heavy hyperparameter-optimimization for this model.
3. Yes! This is expected. It happens exactly because of what you say: some combinations of concepts never appear in the dataset. CelebA is far from unbiased and has a lot of skewed attribute distributions, especially when you look at these concepts jointly.

Please let me know if this was of any help and/or if I can help by answering any further questions. If this looks good to you, then feel free to close the ticket too :) Thank you so much and happy researching!

**Comment by nishadsinghi** [comment]
Hey Mateo, 

Thanks for your response -- it made things clearer! I'll run some follow-up experiments as you suggested and get back to you :)

Best,
Nishad
